lib/asset/AssetModulesPlugin.js

```diff
@@ -101,15 +101,28 @@ class AssetModulesPlugin {
 
 							const AssetGenerator = getAssetGenerator();
 
-							return new AssetGenerator(compilation, dataUrl);
+							return new AssetGenerator(
+								compilation,
+								dataUrl,
+								type === "asset" ? generatorOptions.filename : null
```

Comment:
It makes sense that `filename` is not possible for `asset/inline`, but this should be reflected by the schema validation.

I think you can modify the schema to get all combinations:

``` js
const getGeneratorSchema = memorize(() =>
  require("../../schemas/plugins/AssetModulesPluginGenerator.json")
);
const getGeneratorSchemaForInline = memorize(() => {
  const base = getGeneratorSchema();
  return {
    ...base,
    properties: {
      ...base.properties,
      filename: undefined
    }
  };
}
const getGeneratorSchemaForResource = memorize(() => {
  const base = getGeneratorSchema();
  return {
    ...base,
    properties: {
      ...base.properties,
      dataUrl: undefined
    }
  };
}
```